File: rig4j/src/test/java/com/alflabs/rig4/BlobStoreTest.java
package com.alflabs.rig4;

import com.alflabs.rig4.flags.Flags;
import com.alflabs.utils.FakeFileOps;
import com.alflabs.utils.FileOps;
import com.alflabs.utils.ILogger;
import com.alflabs.utils.MockClock;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

import static com.google.common.truth.Truth.assertThat;

public class BlobStoreTest {
    public @Rule MockitoRule mMockitoRule = MockitoJUnit.rule();

    @Mock private Flags mFlags;
    @Mock private ILogger mLogger;
    private final Timing mTiming = new Timing(new MockClock(), mLogger);

    private FileOps mFileOps;
    private BlobStore mStore;


    @Before
    public void setUp() throws Exception {
        mFileOps = new FakeFileOps();
        mStore = new BlobStore(mFlags, mFileOps, mTiming, mLogger);
    }

    @Test
    public void testString() throws Exception {
        assertThat(mStore.getString("key")).isNull();

        mStore.putString("key", "content");
        assertThat(mStore.getString("key")).isEqualTo("content");
    }

    @Test
    public void testBytes() throws Exception {
        byte[] actual = new byte[]{1, 2, 3, 4};;
        assertThat(mStore.getBytes("key")).isNull();

        mStore.putBytes("key", actual);
        assertThat(mStore.getBytes("key")).isEqualTo(actual);
    }

    @Test
    public void testJson() throws Exception {
        JsonStruct actual = new JsonStruct("The answer is", 42);
        assertThat(mStore.getJson("key", JsonStruct.class)).isNull();

        mStore.putJson("key", actual);
        assertThat(mStore.getJson("key", JsonStruct.class)).isEqualTo(actual);
    }

    public static class JsonStruct {
        public String fieldA;
        public int fieldB;

        // Jackson can create an instance using the default constructor and setting the fields.
        public JsonStruct() {}

        public JsonStruct(String fieldA, int fieldB) {
            this.fieldA = fieldA;
            this.fieldB = fieldB;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            JsonStruct that = (JsonStruct) o;

            if (fieldB != that.fieldB) return false;
            return fieldA != null ? fieldA.equals(that.fieldA) : that.fieldA == null;
        }

        @Override
        public int hashCode() {
            int result = fieldA != null ? fieldA.hashCode() : 0;
            result = 31 * result + fieldB;
            return result;
        }
    }
}

File: rig4k/src/test/kotlin/com/alflabs/rig4k/common/BlobStoreTest.kt
package com.alflabs.rig4k.common

import com.alflabs.utils.FakeFileOps
import com.alflabs.utils.ILogger
import com.alflabs.utils.MockClock
import com.google.common.truth.Truth
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.mockito.Mockito.mock
import org.mockito.junit.MockitoJUnit
import org.mockito.junit.MockitoRule

class BlobStoreTest {
    @get:Rule var mockitoRule: MockitoRule = MockitoJUnit.rule()

    private val logger : ILogger = mock()
    private val options : BlobStoreOptions = mock()
    private val timing = Timing(MockClock(), logger)
    private val fileOps = FakeFileOps()
    private lateinit var mStore: BlobStore

    @Before
    fun setUp() {
        mStore = BlobStore(options, fileOps, logger, timing)
    }

    @Test
    fun testString() {
        Truth.assertThat(mStore.getString("key")).isNull()
        mStore.putString("key", "content")
        Truth.assertThat(mStore.getString("key")).isEqualTo("content")
    }

    @Test
    fun testBytes() {
        val actual = byteArrayOf(1, 2, 3, 4)
        Truth.assertThat(mStore.getBytes("key")).isNull()
        mStore.putBytes("key", actual)
        Truth.assertThat(mStore.getBytes("key")).isEqualTo(actual)
    }

    @Test
    fun testJson() {
        val actual = JsonStruct("The answer is", 42)
        Truth.assertThat(mStore.getJson("key", JsonStruct::class.java)).isNull()
        mStore.putJson("key", actual)
        Truth.assertThat(mStore.getJson("key", JsonStruct::class.java)).isEqualTo(actual)
    }

    class JsonStruct {
        var fieldA: String? = null
        var fieldB = 0

        // Jackson can create an instance using the default constructor and setting the fields.
        constructor()

        constructor(fieldA: String?, fieldB: Int) {
            this.fieldA = fieldA
            this.fieldB = fieldB
        }

        override fun equals(o: Any?): Boolean {
            if (this === o) return true
            if (o == null || javaClass != o.javaClass) return false
            val that = o as JsonStruct
            if (fieldB != that.fieldB) return false
            return if (fieldA != null) fieldA == that.fieldA else that.fieldA == null
        }

        override fun hashCode(): Int {
            var result = if (fieldA != null) fieldA.hashCode() else 0
            result = 31 * result + fieldB
            return result
        }
    }
}
